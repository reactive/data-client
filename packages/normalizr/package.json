{
  "name": "@data-client/normalizr",
  "version": "0.14.17",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://dataclient.io/docs/concepts/normalization",
  "keywords": [
    "normalize",
    "denormalize",
    "flux",
    "redux",
    "data",
    "model",
    "cache",
    "api",
    "memoize",
    "react-native",
    "ios",
    "android",
    "web",
    "schema",
    "typescript",
    "validation",
    "relational",
    "reactive",
    "json"
  ],
  "sideEffects": false,
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=5.4": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.1": {
      "": [
        "ts4.1/index.d.ts"
      ],
      "*": [
        "ts4.1/index.d.ts"
      ]
    },
    ">=4.0": {
      "": [
        "ts4.0/index.d.ts"
      ],
      "*": [
        "ts4.0/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "node.mjs",
    "legacy",
    "ts4.1",
    "ts4.0",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development yarn g:rollup",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production yarn g:rollup",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean ts3.4 ts4.0 ts4.1",
    "build:legacy-types": "yarn g:legacy-types 4.1 4.0 3.4",
    "prepublishOnly": "run build"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/normalizr"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "core-js-pure": "^3.40.0"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/endpoint": "workspace:*",
    "@js-temporal/polyfill": "^0.4.4",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "immutable": "4.2.2",
    "rollup-plugins": "workspace:*"
  }
}
