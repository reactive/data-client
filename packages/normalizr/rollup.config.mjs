import {
  babel,
  commonjs,
  filesize,
  dts,
  resolve,
  terser,
  onwarn,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies).filter(
  dep => !['@babel/runtime'].includes(dep),
);

const name = 'normalizr';

const isProduction = process.env.NODE_ENV === 'production';

const destBase = 'dist/normalizr';
const destExtension = `${isProduction ? '.min' : ''}.js`;
const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];

function snakeCase(id) {
  return id.replace(/(-|\/)/g, '_').replace(/@/g, '');
}

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];

if (process.env.BROWSERSLIST_ENV !== 'node12') {
  if (process.env.NODE_ENV === 'production') {
    configs.push({
      input: 'src/index.ts',
      output: [{ file: `${destBase}.es${destExtension}`, format: 'es' }],
      external: id => id === '..' || isExternal(id),
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          babelHelpers: 'runtime',
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
  } else {
    configs.push({
      input: 'src/index.ts',
      output: [
        {
          file: `${destBase}.umd${destExtension}`,
          format: 'umd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.amd${destExtension}`,
          format: 'amd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.browser${destExtension}`,
          format: 'iife',
          name: snakeCase(name),
        },
      ],
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          babelHelpers: 'runtime',
          rootMode: 'upward',
          caller: { polyfillMethod: false },
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
    configs.push({
      input: './lib/index.d.ts',
      output: [{ file: 'index.d.ts', format: 'es' }],
      plugins: [dts()],
    });
  }
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'src/index.ts',
    output: [{ file: `${destBase}${destExtension}`, format: 'cjs' }],
    external: id => id === '..' || isExternal(id),
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        babelHelpers: 'runtime',
        rootMode: 'upward',
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      isProduction && terser(),
      filesize(),
    ].filter(Boolean),
  });
}
export default configs;
