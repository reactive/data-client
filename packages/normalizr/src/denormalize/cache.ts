import type { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default interface Cache {
  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID;
  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  };
}
