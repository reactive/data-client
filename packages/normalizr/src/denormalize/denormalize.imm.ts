import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';
import { MemoPolicy } from '../memo/Policy.imm.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    MemoPolicy.getEntities(entities),
    new LocalCache(),
    args,
  )(schema, input).data;
}
