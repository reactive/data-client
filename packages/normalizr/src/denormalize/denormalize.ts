import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
