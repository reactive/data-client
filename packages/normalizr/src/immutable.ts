export { DelegateImmutable } from './memo/Delegate.immutable.js';
