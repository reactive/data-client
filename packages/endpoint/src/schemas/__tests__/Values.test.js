// eslint-env jest
import { normalize } from '@data-client/normalizr';
import { IDEntity } from '__tests__/new';
import { fromJS } from 'immutable';

import { SimpleMemoCache, fromJSEntities } from './denormalize';
import { schema } from '../../';
import { INVALID } from '../../special';
import Entity from '../Entity';

let dateSpy;
beforeAll(() => {
  dateSpy = jest

    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

class Cat extends IDEntity {
  type = '';
}
class Dog extends IDEntity {
  type = '';
}

describe(`${schema.Values.name} normalization`, () => {
  let warnSpy;
  afterEach(() => {
    warnSpy.mockRestore();
  });
  beforeEach(() =>
    (warnSpy = jest.spyOn(console, 'warn')).mockImplementation(() => {}),
  );

  test('normalizes without schemaAttribute', () => {
    class MyEntity extends IDEntity {
      name = '';
    }
    const valuesSchema = new schema.Values(MyEntity);

    expect(
      normalize(valuesSchema, {
        first: {
          id: '1',
          name: 'first thing',
        },
        second: {
          id: '2',
          name: 'second thing',
        },
      }),
    ).toMatchSnapshot();
  });

  test('normalizes the values of an object with the given schema', () => {
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      entity => entity.type,
    );

    expect(
      normalize(valuesSchema, {
        fido: { id: '1', type: 'dogs' },
        fluffy: { id: '1', type: 'cats' },
      }),
    ).toMatchSnapshot();
  });

  test('can use a function to determine the schema when normalizing', () => {
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      entity => `${entity.type}s`,
    );

    expect(
      normalize(valuesSchema, {
        fido: { id: '1', type: 'dog' },
        fluffy: { id: '1', type: 'cat' },
        jim: { id: '2', type: 'lizard' },
      }),
    ).toMatchSnapshot();
    expect(warnSpy.mock.calls).toMatchSnapshot();
  });

  test('filters out null and undefined values', () => {
    const valuesSchema = new schema.Values(
      {
        dogs: Dog,
        cats: Cat,
      },
      entity => entity.type,
    );

    expect(
      normalize(valuesSchema, {
        fido: undefined,
        milo: null,
        fluffy: { id: '1', type: 'cats' },
      }),
    ).toMatchSnapshot();
  });

  test('works on complex object', () => {
    class Estimate extends Entity {
      pk() {
        return this.fee.currency;
      }
    }
    const response = {
      data: {
        estimates: {
          BTC: {
            confirmation_duration: 900,
            exchange: {
              rate: '6820.07',
              local: 'USD',
              crypto: 'BTC',
            },
            fee_per_kb: {
              amount: '0.00016566',
              currency: 'BTC',
            },
            fee: {
              amount: '0.00002270',
              currency: 'BTC',
            },
            priority: 'fast',
            recipient_value: {
              amount: '0.00054147',
              currency: 'BTC',
            },
            exchange_to_proceeds: {
              rate: '6820.07',
              local: 'EUR',
              crypto: 'BTC',
            },
            min_order_size: {
              amount: '0.001',
              currency: 'BTC',
            },
            coinbase_fees: {
              amount: '0.00002270',
              currency: 'BTC',
            },
          },
          ETH: {
            confirmation_duration: 900,
            exchange: {
              rate: '197.07',
              local: 'USD',
              crypto: 'ETH',
            },
            fee_per_kb: {
              amount: '0.00086',
              currency: 'ETH',
            },
            fee: {
              amount: '0.03795',
              currency: 'ETH',
            },
            priority: 'fast',
            recipient_value: {
              amount: '2.53',
              currency: 'ETH',
            },
            exchange_to_proceeds: {
              rate: '6820.07',
              local: 'EUR',
              crypto: 'BTC',
            },
            min_order_size: {
              amount: '0.001',
              currency: 'BTC',
            },
            coinbase_fees: {
              amount: '0.00002270',
              currency: 'BTC',
            },
          },
        },
      },
    };
    expect(
      normalize({ data: { estimates: new schema.Values(Estimate) } }, response),
    ).toMatchSnapshot();
  });
});

describe.each([
  ['direct', data => data],
  ['immutable', fromJSEntities],
])(`input (%s)`, (_, createInput) => {
  describe.each([['current', new SimpleMemoCache().denormalize]])(
    `${schema.Values.name} denormalization (%s)`,
    (_, denormalize) => {
      test('denormalizes without schemaAttribute', () => {
        class MyEntity extends IDEntity {
          name = '';
        }
        const valuesSchema = new schema.Values(MyEntity);

        const entities = {
          MyEntity: {
            1: {
              id: '1',
              name: 'first thing',
            },
            2: {
              id: '2',
              name: 'second thing',
            },
          },
        };

        expect(
          denormalize(
            valuesSchema,
            {
              first: '1',
              second: '2',
            },
            createInput(entities),
          ),
        ).toMatchSnapshot();
      });

      test('denormalizes the values of an object with the given schema', () => {
        const valuesSchema = new schema.Values(
          {
            dogs: Dog,
            cats: Cat,
          },
          entity => entity.type,
        );

        const entities = {
          Cat: { 1: { id: '1', type: 'cats' } },
          Dog: { 1: { id: '1', type: 'dogs' } },
        };

        expect(
          denormalize(
            valuesSchema,
            {
              fido: { id: '1', schema: 'dogs' },
              fluffy: { id: '1', schema: 'cats' },
            },
            createInput(entities),
          ),
        ).toMatchSnapshot();
      });

      test('denormalizes with missing entity should have false second value', () => {
        const valuesSchema = new schema.Values(
          {
            dogs: Dog,
            cats: Cat,
          },
          entity => entity.type,
        );

        const entities = {
          Cat: { 1: { id: '1', type: 'cats' } },
          Dog: { 1: { id: '1', type: 'dogs' } },
        };

        expect(
          denormalize(
            valuesSchema,
            {
              fido: { id: '1', schema: 'dogs' },
              fluffy: { id: '1', schema: 'cats' },
              prancy: { id: '5', schema: 'cats' },
            },
            createInput(entities),
          ),
        ).toMatchSnapshot();
      });

      test('denormalizes with deleted entity should just remove them', () => {
        const valuesSchema = new schema.Values(
          {
            dogs: Dog,
            cats: Cat,
          },
          entity => entity.type,
        );

        const entities = {
          Cat: { 1: { id: '1', type: 'cats' }, 5: INVALID },
          Dog: { 1: { id: '1', type: 'dogs' } },
        };

        expect(
          denormalize(
            valuesSchema,
            {
              fido: { id: '1', schema: 'dogs' },
              fluffy: { id: '1', schema: 'cats' },
              prancy: { id: '5', schema: 'cats' },
            },
            createInput(entities),
          ),
        ).toMatchSnapshot();
      });

      test('works on complex object', () => {
        class Estimate extends Entity {
          pk() {
            return this.fee.currency;
          }
        }
        const response = {
          data: {
            estimates: {
              BTC: {
                confirmation_duration: 900,
                exchange: {
                  rate: '6820.07',
                  local: 'USD',
                  crypto: 'BTC',
                },
                fee_per_kb: {
                  amount: '0.00016566',
                  currency: 'BTC',
                },
                fee: {
                  amount: '0.00002270',
                  currency: 'BTC',
                },
                priority: 'fast',
                recipient_value: {
                  amount: '0.00054147',
                  currency: 'BTC',
                },
                exchange_to_proceeds: {
                  rate: '6820.07',
                  local: 'EUR',
                  crypto: 'BTC',
                },
                min_order_size: {
                  amount: '0.001',
                  currency: 'BTC',
                },
                coinbase_fees: {
                  amount: '0.00002270',
                  currency: 'BTC',
                },
              },
              ETH: {
                confirmation_duration: 900,
                exchange: {
                  rate: '197.07',
                  local: 'USD',
                  crypto: 'ETH',
                },
                fee_per_kb: {
                  amount: '0.00086',
                  currency: 'ETH',
                },
                fee: {
                  amount: '0.03795',
                  currency: 'ETH',
                },
                priority: 'fast',
                recipient_value: {
                  amount: '2.53',
                  currency: 'ETH',
                },
                exchange_to_proceeds: {
                  rate: '6820.07',
                  local: 'EUR',
                  crypto: 'BTC',
                },
                min_order_size: {
                  amount: '0.001',
                  currency: 'BTC',
                },
                coinbase_fees: {
                  amount: '0.00002270',
                  currency: 'BTC',
                },
              },
            },
          },
        };
        const shape = new schema.Object({
          data: new schema.Object({ estimates: new schema.Values(Estimate) }),
        });
        const { result, entities } = normalize(shape, response);
        expect(
          denormalize(shape, result, createInput(entities)),
        ).toMatchSnapshot();
      });
    },
  );
});
