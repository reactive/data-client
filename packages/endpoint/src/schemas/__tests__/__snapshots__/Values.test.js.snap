// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValuesSchema normalization can use a function to determine the schema when normalizing 1`] = `
{
  "entities": {
    "Cat": {
      "1": {
        "id": "1",
        "type": "cat",
      },
    },
    "Dog": {
      "1": {
        "id": "1",
        "type": "dog",
      },
    },
  },
  "entitiesMeta": {
    "Cat": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Dog": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "fido": {
      "id": "1",
      "schema": "dogs",
    },
    "fluffy": {
      "id": "1",
      "schema": "cats",
    },
    "jim": {
      "id": "2",
      "type": "lizard",
    },
  },
}
`;

exports[`ValuesSchema normalization can use a function to determine the schema when normalizing 2`] = `
[
  [
    "Schema attribute "lizards" is not expected.
Expected one of: "dogs", "cats"

Value: {
  "id": "2",
  "type": "lizard"
}",
  ],
]
`;

exports[`ValuesSchema normalization filters out null and undefined values 1`] = `
{
  "entities": {
    "Cat": {
      "1": {
        "id": "1",
        "type": "cats",
      },
    },
  },
  "entitiesMeta": {
    "Cat": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "fluffy": {
      "id": "1",
      "schema": "cats",
    },
  },
}
`;

exports[`ValuesSchema normalization normalizes the values of an object with the given schema 1`] = `
{
  "entities": {
    "Cat": {
      "1": {
        "id": "1",
        "type": "cats",
      },
    },
    "Dog": {
      "1": {
        "id": "1",
        "type": "dogs",
      },
    },
  },
  "entitiesMeta": {
    "Cat": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Dog": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "fido": {
      "id": "1",
      "schema": "dogs",
    },
    "fluffy": {
      "id": "1",
      "schema": "cats",
    },
  },
}
`;

exports[`ValuesSchema normalization normalizes without schemaAttribute 1`] = `
{
  "entities": {
    "MyEntity": {
      "1": {
        "id": "1",
        "name": "first thing",
      },
      "2": {
        "id": "2",
        "name": "second thing",
      },
    },
  },
  "entitiesMeta": {
    "MyEntity": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "first": "1",
    "second": "2",
  },
}
`;

exports[`ValuesSchema normalization works on complex object 1`] = `
{
  "entities": {
    "Estimate": {
      "BTC": {
        "coinbase_fees": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": {
          "crypto": "BTC",
          "local": "USD",
          "rate": "6820.07",
        },
        "exchange_to_proceeds": {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "fee_per_kb": {
          "amount": "0.00016566",
          "currency": "BTC",
        },
        "min_order_size": {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": {
          "amount": "0.00054147",
          "currency": "BTC",
        },
      },
      "ETH": {
        "coinbase_fees": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": {
          "crypto": "ETH",
          "local": "USD",
          "rate": "197.07",
        },
        "exchange_to_proceeds": {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": {
          "amount": "0.03795",
          "currency": "ETH",
        },
        "fee_per_kb": {
          "amount": "0.00086",
          "currency": "ETH",
        },
        "min_order_size": {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": {
          "amount": "2.53",
          "currency": "ETH",
        },
      },
    },
  },
  "entitiesMeta": {
    "Estimate": {
      "BTC": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "ETH": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "data": {
      "estimates": {
        "BTC": "BTC",
        "ETH": "ETH",
      },
    },
  },
}
`;

exports[`input (direct) ValuesSchema denormalization (current) denormalizes the values of an object with the given schema 1`] = `
{
  "fido": Dog {
    "id": "1",
    "type": "dogs",
  },
  "fluffy": Cat {
    "id": "1",
    "type": "cats",
  },
}
`;

exports[`input (direct) ValuesSchema denormalization (current) denormalizes with deleted entity should just remove them 1`] = `
{
  "fido": Dog {
    "id": "1",
    "type": "dogs",
  },
  "fluffy": Cat {
    "id": "1",
    "type": "cats",
  },
}
`;

exports[`input (direct) ValuesSchema denormalization (current) denormalizes with missing entity should have false second value 1`] = `
{
  "fido": Dog {
    "id": "1",
    "type": "dogs",
  },
  "fluffy": Cat {
    "id": "1",
    "type": "cats",
  },
}
`;

exports[`input (direct) ValuesSchema denormalization (current) denormalizes without schemaAttribute 1`] = `
{
  "first": MyEntity {
    "id": "1",
    "name": "first thing",
  },
  "second": MyEntity {
    "id": "2",
    "name": "second thing",
  },
}
`;

exports[`input (direct) ValuesSchema denormalization (current) works on complex object 1`] = `
{
  "data": {
    "estimates": {
      "BTC": Estimate {
        "coinbase_fees": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": {
          "crypto": "BTC",
          "local": "USD",
          "rate": "6820.07",
        },
        "exchange_to_proceeds": {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "fee_per_kb": {
          "amount": "0.00016566",
          "currency": "BTC",
        },
        "min_order_size": {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": {
          "amount": "0.00054147",
          "currency": "BTC",
        },
      },
      "ETH": Estimate {
        "coinbase_fees": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": {
          "crypto": "ETH",
          "local": "USD",
          "rate": "197.07",
        },
        "exchange_to_proceeds": {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": {
          "amount": "0.03795",
          "currency": "ETH",
        },
        "fee_per_kb": {
          "amount": "0.00086",
          "currency": "ETH",
        },
        "min_order_size": {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": {
          "amount": "2.53",
          "currency": "ETH",
        },
      },
    },
  },
}
`;

exports[`input (immutable) ValuesSchema denormalization (current) denormalizes the values of an object with the given schema 1`] = `
{
  "fido": Dog {
    "id": "1",
    "type": "dogs",
  },
  "fluffy": Cat {
    "id": "1",
    "type": "cats",
  },
}
`;

exports[`input (immutable) ValuesSchema denormalization (current) denormalizes with deleted entity should just remove them 1`] = `
{
  "fido": Dog {
    "id": "1",
    "type": "dogs",
  },
  "fluffy": Cat {
    "id": "1",
    "type": "cats",
  },
}
`;

exports[`input (immutable) ValuesSchema denormalization (current) denormalizes with missing entity should have false second value 1`] = `
{
  "fido": Dog {
    "id": "1",
    "type": "dogs",
  },
  "fluffy": Cat {
    "id": "1",
    "type": "cats",
  },
}
`;

exports[`input (immutable) ValuesSchema denormalization (current) denormalizes without schemaAttribute 1`] = `
{
  "first": MyEntity {
    "id": "1",
    "name": "first thing",
  },
  "second": MyEntity {
    "id": "2",
    "name": "second thing",
  },
}
`;

exports[`input (immutable) ValuesSchema denormalization (current) works on complex object 1`] = `
{
  "data": {
    "estimates": {
      "BTC": Estimate {
        "coinbase_fees": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": {
          "crypto": "BTC",
          "local": "USD",
          "rate": "6820.07",
        },
        "exchange_to_proceeds": {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "fee_per_kb": {
          "amount": "0.00016566",
          "currency": "BTC",
        },
        "min_order_size": {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": {
          "amount": "0.00054147",
          "currency": "BTC",
        },
      },
      "ETH": Estimate {
        "coinbase_fees": {
          "amount": "0.00002270",
          "currency": "BTC",
        },
        "confirmation_duration": 900,
        "exchange": {
          "crypto": "ETH",
          "local": "USD",
          "rate": "197.07",
        },
        "exchange_to_proceeds": {
          "crypto": "BTC",
          "local": "EUR",
          "rate": "6820.07",
        },
        "fee": {
          "amount": "0.03795",
          "currency": "ETH",
        },
        "fee_per_kb": {
          "amount": "0.00086",
          "currency": "ETH",
        },
        "min_order_size": {
          "amount": "0.001",
          "currency": "BTC",
        },
        "priority": "fast",
        "recipient_value": {
          "amount": "2.53",
          "currency": "ETH",
        },
      },
    },
  },
}
`;
